drivers/d3d12/rendering_device_driver_d3d12.cpp

```diff
@@ -319,56 +319,203 @@ const RenderingDeviceDriverD3D12::D3D12Format RenderingDeviceDriverD3D12::RD_TO_
+	HRESULT res = device->CreateDescriptorHeap(&desc, IID_PPV_ARGS(&fallback_heap));
+
+	ERR_FAIL_COND_MSG(!SUCCEEDED(res), "CreateDescriptorHeap failed with error " + vformat("0x%08ux", (uint64_t)res) + ".");
+	fallback = { fallback_heap, 0, 0, false, this };
+	storage.push_back(fallback);
+}
+
+Error RenderingDeviceDriverD3D12::CPUDescriptorsHeapPool::allocate(CPUDescriptorsHeapHandle &result) {
+	if (failed_last_alloc) {
+		result = fallback;
+		return OK;
+	}
+
+	if (free_stack.size() == 0) {
+		D3D12_DESCRIPTOR_HEAP_DESC descr_copy = desc;
+		// allow for more than 1024 descriptors
+		descr_copy.NumDescriptors = MAX(desc.NumDescriptors, 1024u);
+
+		ID3D12DescriptorHeap *heap;
+		HRESULT res = device->CreateDescriptorHeap(&descr_copy, IID_PPV_ARGS(&heap));
+		if (!SUCCEEDED(res)) {
+			_err_print_error(FUNCTION_STR, __FILE__, __LINE__, "CreateDescriptorHeap failed with error " + vformat("0x%08ux", (uint64_t)res) + ".");
+			result = fallback;
+			failed_last_alloc = true;
+			return OK;
+		}
+
+		// fill storage in inverse order
+
+		for (int i = descr_copy.NumDescriptors - desc.NumDescriptors; i >= 0; i -= desc.NumDescriptors) {
+			uint32_t storage_index = storage.size();
+			CPUDescriptorsHeapHandle handle = { heap, (uint32_t)i, storage_index, true, this };
+			storage.push_back(handle);
+			free_stack.push_back(storage_index);
+		}
+	}
+
+	int free_stack_index = free_stack.size() - 1;
+	int index = free_stack[free_stack_index];
+	free_stack.resize(free_stack_index);
+	CPUDescriptorsHeapHandle &storage_item = storage[index];
+	if (!storage_item.disposed) {
+		ERR_FAIL_COND_V_MSG(!storage_item.disposed, ERR_INVALID_DATA, "CPUDescriptorsHeapPool is in broken state!");
+	}
+	storage_item.disposed = false;
+	result = storage_item;
+	return OK;
+}
+Error RenderingDeviceDriverD3D12::CPUDescriptorsHeapPool::release(const CPUDescriptorsHeapHandle &p_handle) {
+	if (p_handle.storage_index == 0) { // cannot free fallback
+		return OK;
+	}
+	bool &disposed = storage[p_handle.storage_index].disposed;
+	ERR_FAIL_COND_V_MSG(disposed, ERR_INVALID_PARAMETER, "CPUDescriptorsHeapHandle double free!");
+	disposed = true;
+	free_stack.push_back(p_handle.storage_index);
+	failed_last_alloc = false;
+	return OK;
+}
+
+static uint32_t to_next_pow2(uint32_t val) {
+	val--;
+	val |= val >> 1;
+	val |= val >> 2;
+	val |= val >> 4;
+	val |= val >> 8;
+	val |= val >> 16;
+	val++;
+	return val;
+}
+Error RenderingDeviceDriverD3D12::CPUDescriptorsHeapPools::allocate(ID3D12Device *p_device, const D3D12_DESCRIPTOR_HEAP_DESC &p_desc, CPUDescriptorsHeapHandle &result) {
+	ERR_FAIL_COND_V(p_desc.NodeMask != 0 || p_desc.Flags != 0, ERR_INVALID_PARAMETER);
+
+	// bin number of descriptors to the next power of two
+	uint32_t num_rounded_pow2 = to_next_pow2(p_desc.NumDescriptors);
+	uint32_t hash_table_key = num_rounded_pow2 * D3D12_DESCRIPTOR_HEAP_TYPE_NUM_TYPES;
+	hash_table_key += p_desc.Type;
+
+	// is locking needed here ?
+	MutexLock lock(mutex);
```

Comment:
Presumably it's not, but only because RenderingDevice itself declares uniform set creation as a thread safe method, something that may very well go away in the future. I'd keep it.